### ISSUE TITLE
Stop using abbreviations for Whonix templates

### ISSUE THREAD

**Issue by mfc** [QubesOS/qubes-issues#1778]
we need to stop using non-intuitive abbreviations, which are:
- confusing for new users
- bad for localization/translation (directly or in documentation)

`whonix-ws` -> `whonix-workstation`
`whonix-gw` -> `whonix-gateway`

**Comment by marmarek** [comment]
I don't think it is great idea. Those VMs are not actually
whonix-workstation or whonix-gateway. Those are templates of them. It
requires more thinking. Maybe templates should be hidden by default
(renamed by your suggestion)? But then there should be easy and
intuitive way for doing updates and installing new software...

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by mfc** [comment]
I agree the proposed names are bad, they should highlight that they are templates (maybe `whonix-workstation-template` etc?). What about the underlying idea of reducing use of abbreviations?

**Comment by adrelanos** [comment]
Perhaps all TemplateVMs should be prefixed with 'template-'?

Looping in @bnvk. How does this ticket integrate with what you had in
mind for the QVMM revamp / alternative gui?

**Comment by axon-qubes** [comment]
I like @adrelanos's suggestion of prefixing all TemplateVMs. (Personally, I would prefer something like `tmpl-` over `template-` because my template names are already too long.)

Prefixing is much better than postfixing in this case, because it will result in all the templates sorting together lexically.

**Comment by mfc** [comment]
> (Personally, I would prefer something like tmpl- over template- because my template names are already too long.)

No, [no abbreviations](https://github.com/QubesOS/qubes-issues/issues/1776). We should have tab-complete to make typing such things in terminal faster. Or have no abbreviations and have clearer separation of different types of VMs (template, system, app) clearer in the VM Manager.

> I don't think it is great idea. Those VMs are not actually whonix-workstation or whonix-gateway. Those are templates of them.

You can't argue `whonix-ws` -> `whonix-workstation` is bad because people will think they are not templates, that is an orthogonal problem to this and is a problem with all templates. e.g. if that is a problem, it already exists, and this does not address that problem.

I continue to run into issues with these abbreviations at trainings, whether they are native english speakers or not. The worst part is that `whonix-ws` and `whonix-gw` are only one letter apart >:(

**Comment by adrelanos** [comment]
What's the status of this ticket?

@marmarek 
> I don't think it is great idea.

At this point, considered rejected as in "won't fix" or "won't do"?

Or waiting for more discussion / thinking / input?

Or is it dependent on the outcome of https://github.com/QubesOS/qubes-issues/issues/1776?

----

(I personally don't have any strong opinion here either way and would to refer to advice by usability designers.)

**Comment by DemiMarie** [comment]
@ninavizz I imagine this is your call

**Comment by ninavizz** [comment]
Yes @adrelanos I would qualify this as dependent on #1776. HOWEVER, I also see #1776 as nearing completion. 

For the latest Whonix release, I'd prefer we not change anything, in the interest of not creating UX churn. For `4.2` I see #1776 being resolved, as the new app menu design already does that (and I suspect Qubes Manager can be tweaked to meet the needs of #1776 by then, too).

**Comment by andrewdavidwong** [comment]
Isn't this just a specific instance of #1776? What's the point of having both open?

**Comment by mfc** [comment]
I don't see this issue at all dependent to #1776. 

the equivalent of this issue is if we named the debian or fedora minimal templates to be `debian-11-mn` or `fedora-11-ml`. this issue is focused on a behavior that is not system-wide, it is specific to the whonix templates and has no logical consistency with the rest of our namings of templates.

all this issue is asking, is  -- just as we spell out `minimal` for the minimal templates -- to spell out what the different whonix templates are. so expand "gw" to "gateway" and "ws" to "workstation". that is all.

this can be done independently of #1776 and we could even not implement #1776 and still fix this issue.

**Comment by ninavizz** [comment]
@mfc Ok, so—two things. One, your point about localization and multi-lingual users is totally legit. "disp-89068" as an example, makes no sense to users that don't speak English, or do speak English but don't quickly intuit that "disp" is an abbreviation of "disposable." 

With `4.1` we do have clearer icons to delineate what is a template, vs what is disposable, vs what is an app qube, vs what is a service qube. And yet Whonix is... (winces) a Service that requires Templates, and may not even be seen as a Service by the system. BUT, Whonix Templates will still get Template icons per #5676 

For users using the new app menu, App qubes and Templates will further never be shown in the same view, which will help keep users from going to open a Template when they want to open an App qube. 

Given the icons will help disambiguate App qubes from Templates, would you feel satisfied with `whonix-gateway` and `whonix-workstation` as qube names for Whonix 16?

**Comment by ninavizz** [comment]
@andrewdavidwong I think this issue has pivoted to discussing Template name suffixes, not prefixes; but the abbreviations issue cited in the title, began as a discussion around prefixes. Both stink, tbh; but also need a longer-term solution (like a parenthetical or something). I dunno. It still feels like this is not yet solved for.

**Comment by mfc** [comment]
correct this issue is not about the prefixes and is not something that icons or ordering can help solve -- all of that is in #1776. 

if we follow the [naming convention of the minimal templates](https://www.qubes-os.org/doc/templates/minimal/) which is `qubes-template-<DISTRO_NAME>-<RELEASE_NUMBER>-minimal` then the upcoming Whonix 16 templates should be named in the repos:
* `qubes-template-whonix-16-workstation`
* `qubes-template-whonix-16-gateway`

which would appear when installed to the user as:
* `whonix-16-workstation`
* `whonix-16-gateway`

or if we consider the workstation and gateway to be separate distributions (our current practice, i'm not sure why), then we can switch the release number and workstation/gateway as we currently do. but i think the above is cleaner & clearer.

**Comment by andrewdavidwong** [comment]
Thanks for the explanation. I understand now.

**Comment by marmarek** [comment]
Ok, indeed with clearer distinction what is a template and what is not (new menu, new icons, new layout in the manager etc), I think we can go ahead with this rename.
Some things to keep in mind when choosing names are:
- qube name needs to be system-wide unique (you can't have App qube and Template qube with the same name)
- qube names are limited to 31 characters

But in this case, none of the above is a concern. Should be good.

@adrelanos do you want to go with this with Whonix 16 already, or is it too late for such change, and should we schedule it for Whonix 17?

**Comment by adrelanos** [comment]
> correct this issue is not about the prefixes and is not something that icons or ordering can help solve -- all of that is in #1776.
> 
> if we follow the [naming convention of the minimal templates](https://www.qubes-os.org/doc/templates/minimal/) which is `qubes-template-<DISTRO_NAME>-<RELEASE_NUMBER>-minimal` then the upcoming Whonix 16 templates should be named in the repos:
> 
> * `qubes-template-whonix-16-workstation`
> * `qubes-template-whonix-16-gateway`
> 
> which would appear when installed to the user as:
> 
> * `whonix-16-workstation`
> * `whonix-16-gateway`

I would prefer the long notation `qubes-template-whonix-16-workstation`, `qubes-template-whonix-16-gateway` for all places.

And I'd suggest to avoid the short notation `whonix-16-workstation`, `whonix-16-gateway` for templates. Any icons, tables are useful but for blind users or otherwise people who easily overlook such things or who mostly rely on textual information that notation can be confusing. While the long notation is unambiguous.

...but I am missing QVMM R4.1 innovations. Maybe hard to get it wrong there.

> @adrelanos do you want to go with this with Whonix 16 already, or is it too late for such change, and should we schedule it for Whonix 17?

Might cause confusion if some have Whonix 16 in the legacy notation and some not if we change that now. Not sure.

Doing this for Whonix 17 seems safe. Pull requests would be useful with a note to merge these for Whonix 17 so this doesn't have to be invented under pressure when Whonix 17 release time comes close or gets forgotten by then.

**Comment by andrewdavidwong** [comment]
> > correct this issue is not about the prefixes and is not something that icons or ordering can help solve -- all of that is in #1776.
> 
> > 
> 
> > if we follow the [naming convention of the minimal templates](https://www.qubes-os.org/doc/templates/minimal/) which is `qubes-template-<DISTRO_NAME>-<RELEASE_NUMBER>-minimal` then the upcoming Whonix 16 templates should be named in the repos:
> 
> > 
> 
> > * `qubes-template-whonix-16-workstation`
> 
> > * `qubes-template-whonix-16-gateway`
> 
> > 
> 
> > which would appear when installed to the user as:
> 
> > 
> 
> > * `whonix-16-workstation`
> 
> > * `whonix-16-gateway`
> 
> 
> 
> I would prefer the long notation `qubes-template-whonix-16-workstation`, `qubes-template-whonix-16-gateway` for all places.
> 

That's the naming pattern for template *packages*, not qubes.

Now you're talking about changing the naming pattern for *all* template qubes, not just Whonix templates.

Also, the first name is longer than 31 characters.

> 
> And I'd suggest to avoid the short notation `whonix-16-workstation`, `whonix-16-gateway` for templates. Any icons, tables are useful but for blind users or otherwise people who easily overlook such things or who mostly rely on textual information that notation can be confusing. While the long notation is unambiguous.
> 
> 
> 
> ...but I am missing QVMM R4.1 innovations. Maybe hard to get it wrong there.
> 
> 
> 
> > @adrelanos do you want to go with this with Whonix 16 already, or is it too late for such change, and should we schedule it for Whonix 17?
> 
> 
> 
> Might cause confusion if some have Whonix 16 in the legacy notation and some not if we change that now. Not sure.
> 

But no one has Whonix 16 _stable_ yet, because it doesn't exist yet. Changes to Whonix 16 _unstable_ are fine, as it's unstable. (Note: I'm not saying you should or shouldn't do it. I'm only addressing this specific point.)

> 
> Doing this for Whonix 17 seems safe. Pull requests would be useful with a note to merge these for Whonix 17 so this doesn't have to be invented under pressure when Whonix 17 release time comes close or gets forgotten by then.

**Comment by ninavizz** [comment]
@mfc My comments about icons, are relevant to whether or not the word "template" is included in the name of the qube. TY @andrewdavidwong for clarifying that Michael was in fact describing _packages_ and not _qube_ names. 

Question for everyone: Why is it normal/customary in Qubes to leave old versions of things on a machine, when we have new versions? Like, is there a _reason_ to have multiple versions of Whonix on a machine, other than "It doesn't un-install old versions"?

I'd like to develop some method for tracking distro versions other than forcing them to be included in the name. At some point, not as a part of this ticket. Yes, I'm going to be chatting with @marmarta about including build/version numbers for templates in their submenus from the App Menu, and for Qubes Manager a unique UIs (variation on the existing) will also need to be developed). On the CLI some UX things will likely need to be added.

**Comment by andrewdavidwong** [comment]
> Question for everyone: Why is it normal/customary in Qubes to leave old versions of things on a machine, when we have new versions? Like, is there a reason to have multiple versions of Whonix on a machine, other than "It doesn't un-install old versions"?

Some users like to have multiple templates corresponding to multiple releases of an OS installed at the same time (sometimes even EOL releases). Users have a wide variety of reasons for this practice. It would be very bad if we automatically deleted any of their templates without asking simply because we assumed they were "old," and it would probably fail a lot of the time anyway, since there would probably be app qubes dependent on those templates. However, it might make sense to _prompt_ users to _ask_ them whether they'd like us to delete an old template that we automatically detect is no longer in use (or _inform_ them that the old one is still in use, possibly unintentionally), e.g., as a final cleanup step after installing a new template version.

**Comment by brendanhoar** [comment]
Succinctly: a user can customize a qubes-installed template. Deleting a template loses those customizations.

Customization ranges from the common use case of installing additionally software to...changing system configuration to...a universe of things.

A replacement template ("newly built version" not just updates to existing one) won't have those customizations and automatically deleting an old template loses those customizations.

B

**Comment by andrewdavidwong** [comment]
Yeah. Basically, templates are (or, more precisely, _can be_) user data, and you never mess with user data.

**Comment by mfc** [comment]
what we talking around is that the *process* for upgrading customized templates in Qubes sucks and is very manual. i'm sure there are ways of making it smoother. 

for instance, to be able to get a list what additional packages were installed in a customized template, and what other settings changes were implemented (compared to the non-customized template). could even include a prompt to try installing those same packages or implement those same changes in another (newer OS) template. but even just a list would be helpful for migration.

i created an issue to track, feel free to contribute / discuss further there:

https://github.com/QubesOS/qubes-issues/issues/6904

**Comment by ninavizz** [comment]
@mfc I actually added a note to the Project "Updater Improvements" to describe the broader issue, as I feel it needs to be built into a new Updater experience. 

Back to this issue's topix tho: What do you feel the acceptance criteria should be to get this issue closed? @adrelanos responding with something having been done, or something else?

**Comment by adrelanos** [comment]
@andrewdavidwong 
> > > correct this issue is not about the prefixes and is not something that icons or ordering can help solve -- all of that is in #1776.
> > 
> > 
> > > 
> > 
> > 
> > > if we follow the [naming convention of the minimal templates](https://www.qubes-os.org/doc/templates/minimal/) which is `qubes-template-<DISTRO_NAME>-<RELEASE_NUMBER>-minimal` then the upcoming Whonix 16 templates should be named in the repos:
> > 
> > 
> > > 
> > 
> > 
> > > * `qubes-template-whonix-16-workstation`
> > 
> > 
> > > * `qubes-template-whonix-16-gateway`
> > 
> > 
> > > 
> > 
> > 
> > > which would appear when installed to the user as:
> > 
> > 
> > > 
> > 
> > 
> > > * `whonix-16-workstation`
> > 
> > 
> > > * `whonix-16-gateway`
> > 
> > 
> > I would prefer the long notation `qubes-template-whonix-16-workstation`, `qubes-template-whonix-16-gateway` for all places.
> 
> That's the naming pattern for template _packages_, not qubes.
> 
> Now you're talking about changing the naming pattern for _all_ template qubes, not just Whonix templates.
> 
> Also, the first name is longer than 31 characters.

I'd suggest to lift the hard limit of 31 characters.

> > And I'd suggest to avoid the short notation `whonix-16-workstation`, `whonix-16-gateway` for templates. Any icons, tables are useful but for blind users or otherwise people who easily overlook such things or who mostly rely on textual information that notation can be confusing. While the long notation is unambiguous.
> > ...but I am missing QVMM R4.1 innovations. Maybe hard to get it wrong there.
> > > @adrelanos do you want to go with this with Whonix 16 already, or is it too late for such change, and should we schedule it for Whonix 17?
> > 
> > 
> > Might cause confusion if some have Whonix 16 in the legacy notation and some not if we change that now. Not sure.
> 
> But no one has Whonix 16 _stable_ yet, because it doesn't exist yet. Changes to Whonix 16 _unstable_ are fine, as it's unstable. (Note: I'm not saying you should or shouldn't do it. I'm only addressing this specific point.)

It's in the Qubes stable repo. ([staged rollout](https://forums.whonix.org/t/staged-rollouts-staged-upgrades-phased-updates/10932))

* https://github.com/QubesOS/qubes-issues/issues/6891
* https://github.com/QubesOS/updates-status/issues/2613
* https://github.com/QubesOS/updates-status/issues/2614

@ninavizz 
> Back to this issue's topix tho: What do you feel the acceptance criteria should be to get this issue closed? @adrelanos responding with something having been done, or something else?

I'd appreciate if someone could make a pull request for the template names, note that this would have to be merged for Qubes-Whonix 17.

**Comment by ninavizz** [comment]
I'm personally _not_ in favor of increasing the character limit for Template names. We're making great progress with improvements to other areas of the experience that long-character naming conventions were created to remedy.

@andrewdavidwong Since the naming part of this seems to have been resolved, what do you feel acceptance criteria are for this issue at this point? The PR Patrick pointed me to (that I regrettably lack the skills to make)?

**Comment by adrelanos** [comment]
What's the status of this?
Who would make the pull request?

If it's only `C: builder` it might imply that people are waiting for me to implement this.

For issue tracking, what would be the correct component (`C:`)? `C: builder`?

**Comment by DemiMarie** [comment]
@adrelanos I would imagine this would be your job to make the PR, if for no other reason than to ensure there are no hard-coded dependencies on the old names.

**Comment by adrelanos** [comment]
Could you please confirm that template names shall be changed to:

* `whonix-workstation`
* `whonix-gateway`

? @marmarek

----

As for making the pull request... These are the files I found where `whonix-gw` / `whonix-ws`.

* [1] https://github.com/adrelanos/qubes-template-whonix/blob/master/builder.conf#L63 - this file needs to be updated, trivial change for Whonix 17 (pull request could be easily created)
* [2] https://github.com/QubesOS/qubes-core-admin-addon-whonix/blob/master/qubeswhonix/__init__.py#L30 - but this file probably needs no changes
* [3] Whonix source code: Nowhere it is to hardcoded `whonix-gw` / `whonix-ws`. Fortunately, was never any need to hardcode the name of the templates.

So basically it's just `[1]`.

What are the other places where this needs to be changed? Qubes installer? What else? Usually @marmarek handled that.

I am not sure I'll find all the places where it could be hardcoded or orchestrate a template name change without breaking anything. I can maintain the "inside" of the template but not so much the "outside" / Qubes build system / installer which is 99+% unspecific to Whonix.

**Comment by adrelanos** [comment]
> With a version number added, right?

Yes.

> When do you expect first builds of Whonix 17?

Not before Debian `bookworm` becoming the next Debian stable.

**Comment by marmarta** [comment]
Other places to check: new config tool (desktop-linux-manager) repo.

**Comment by adrelanos** [comment]
The first developers-only Debian `bookworm` based Non-Qubes-Whonix builds have been produced.

Quote @marmarek:

> Yes, I can handle that. Those places are:

Now would be a good time to make this change as it was discussed here in this ticket. @marmarek

**Comment by marmarek** [comment]
Proposed order of actions:
1. Adjust build configs so new names can be used for building Whonix 17
2. Build some testing templates using new names
3. Adjust places using the template names (config tools, installer etc) to match new names (some built templates are necessary to test the change)

Do you want to build any Whonix 17 template for R4.1 even if just for testing? Or can I introduce new naming only in R4.2 configs?

**Comment by adrelanos** [comment]
> Proposed order of actions:
> 
> 1. Adjust build configs so new names can be used for building Whonix 17
> 2. Build some testing templates using new names
> 3. Adjust places using the template names (config tools, installer etc) to match new names (some built templates are necessary to test the change)

Sounds good!

> Do you want to build any Whonix 17 template for R4.1 even if just for testing?

I guess best avoided, lower effort, less complexity?

> Or can I introduce new naming only in R4.2 configs?

I think so.

**Comment by marmarek** [comment]
Reviewing the discussion above, I take it should be `whonix-17-gateway`/`whonix-17-workstation` and not `whonix-gateway-17`/`whonix-workstation-17`, right?

**Comment by adrelanos** [comment]
Thank you for asking!

Actually nowadays I disagree with my former self from 2016. Now I am certain that I prefer:

* `whonix-gateway-17`
* `whonix-workstation-17`

(It lacks the first characters in upper case but that's OK because that's how Qubes is doing it.)

This is because the words `Whonix-Gateway` and `Whonix-Workstation` are consistently used on the Whonix website as a proper nouns. At no point we used "Whonix 16 Gateway" or similar. Always "Whonix-Gateway 16" and so forth.

Adding the version number in the middle would encourage users to coin idiosyncratic terms such as "my gateway" or "my workstation", which can lead to confusion. Whonix hasn't monopolized the terms "gateway" or "workstation". For example, FPF has developed a `SecureDrop Workstation`.

**Comment by adrelanos** [comment]
Since the renamed templates are now in the (testers) repository... Has this been completed?

**Comment by marmarek** [comment]
There is one more place - update template names in the installer (currently it still ships with Whonix 16)

**Comment by adrelanos** [comment]
Qubes dom0 appmenu is empty for the R4.2 testing Template. Any renaming required for that?

**Comment by marmarek** [comment]
I've opened remaining PRs related to template renaming. They are not tested yet, I'll build test installer and try it out.

**Comment by marmarek** [comment]
> Qubes dom0 appmenu is empty for the R4.2 testing Template. Any renaming required for that?

Appmenu should work regardless, something went wrong. Maybe it's about "bullseye" in directory names at https://github.com/whonix/qubes-template-whonix ?

**Comment by adrelanos** [comment]
Also https://github.com/Whonix/qubes-template-whonix/blob/master/builder.conf needs to be updated? I can do that now just in case.

But it seems that's not a big ideal? Is the `builder.conf` no longer required?

**Comment by marmarek** [comment]
> But it seems that's not a big ideal? Is the `builder.conf` no longer required?

I think it's used for manual build using qubes-builder(v1). Neither builderv2 nor official builds with legacy qubes-builder use this file (the latter use https://github.com/QubesOS/qubes-template-configs/blob/master/R4.1/templates-community/whonix-16.conf).

**Comment by adrelanos** [comment]
> > Qubes dom0 appmenu is empty for the R4.2 testing Template. Any renaming required for that?
> 
> Appmenu should work regardless, something went wrong. Maybe it's about "bullseye" in directory names at https://github.com/whonix/qubes-template-whonix ?

Quite likely. It's fixed in git. Will re-check in the next template build once https://github.com/QubesOS/qubes-issues/issues/8330 was resolved (https://github.com/QubesOS/qubes-builder-debian/pull/73).

> > But it seems that's not a big ideal? Is the `builder.conf` no longer required?
> 
> I think it's used for manual build using qubes-builder(v1). Neither builderv2 nor official builds with legacy qubes-builder use this file (the latter use https://github.com/QubesOS/qubes-template-configs/blob/master/R4.1/templates-community/whonix-16.conf).

Alright. I'll remove that file now.

**Comment by marmarek** [comment]
Three PRs linked above are tested now.

**Comment by marmarek** [comment]
This rename is done now in R4.2.
